src/include/processor/warning_context.h

```diff
@@ -19,20 +40,41 @@ struct PopulatedCSVError {
 
 struct WarningInfo {
     uint64_t queryID;
-    processor::PopulatedCSVError warning;
+    PopulatedCSVError warning;
 
-    WarningInfo(processor::PopulatedCSVError warning, uint64_t queryID)
+    WarningInfo(PopulatedCSVError warning, uint64_t queryID)
         : queryID(queryID), warning(std::move(warning)) {}
 };
 
-struct WarningContext {
-    std::vector<WarningInfo> warnings;
-    main::ClientConfig* clientConfig;
-    std::mutex mtx;
+using populate_func_t = std::function<PopulatedCSVError(const SharedWarning*)>;
+using populate_error_func_t =
+    std::function<PopulatedCSVError(const SharedWarning*, uint64_t, uint64_t, uint64_t)>;
+
+struct UnpopulatedWarnings {
+    std::vector<std::shared_ptr<SharedWarning>> warnings;
+};
 
+class WarningContext {
+public:
     explicit WarningContext(main::ClientConfig* clientConfig);
 
-    void appendWarningMessages(const std::vector<PopulatedCSVError>& messages, uint64_t queryID);
+    void appendWarningMessages(const std::vector<std::shared_ptr<SharedWarning>>& messages,
+        const WarningSource& warningSource);
+    void populateWarnings(const WarningSource& warningSource, uint64_t queryID,
+        populate_func_t populateFunc);
+    const std::vector<WarningInfo>& getPopulatedWarnings() const;
```

Comment:
Warnings have the following lifecycle:
1. Reported warning cached in operator's local warning cache in raw form
2. When local warning cache is full, raw warnings are flushed to shared warning cache
3. In finalize stage, raw warnings are converted to populated warnings (which have line numbers, reconstructed lines, etc.)

Currently step 2 is partially done in the `WarningContext` (the local cache directly flushes raw warnings to `WarningContext` which stores both raw and populated warnings separately). This is unnecessary + more complicated so I will move step 2 to be done in operators (cached in shared state) with common code.